blockstore/badger/versions/badger_v2.go

```diff
@@ -0,0 +1,314 @@
+}
+
+func (b *BadgerV2) IndexCacheMetrics() *ristretto.Metrics {
+	return b.DB.IndexCacheMetrics()
+}
+
+func (b *BadgerV2) GetErrKeyNotFound() error {
+	return badger.ErrKeyNotFound
+}
+
+func (b *BadgerV2) GetErrNoRewrite() error {
+	return badger.ErrNoRewrite
+}
+
+func (b *BadgerV2) NewWriteBatch() WriteBatch {
+	return &BadgerV2WriteBatch{b.DB.NewWriteBatch()}
+}
+
+func (b *BadgerV2) Flatten(workers int) error {
+	return b.DB.Flatten(workers)
+}
+
+func (b *BadgerV2) Size() (lsm int64, vlog int64) {
+	return b.DB.Size()
+}
+
+func (b *BadgerV2) Copy(ctx context.Context, to BadgerDB) (defErr error) {
+
+	batch := to.NewWriteBatch()
+	defer func() {
+		if defErr == nil {
+			defErr = batch.Flush()
+		}
+		if defErr != nil {
+			batch.Cancel()
+		}
+	}()
+
+	stream := b.DB.NewStream()
+
+	return iterateBadger(ctx, stream, func(kvs []*pb.KV) error {
+		// check whether context is closed on every kv group
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		for _, kv := range kvs {
+			if err := batch.Set(kv.Key, kv.Value); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+var IterateLSMWorkers int // defaults to between( 2, 8, runtime.NumCPU/2 )
+
+func iterateBadger(ctx context.Context, stream *badger.Stream, iter func([]*pb.KV) error) error {
+	workers := IterateLSMWorkers
+	if workers == 0 {
+		workers = between(2, 8, runtime.NumCPU()/2)
+	}
+
+	stream.NumGo = workers
+	stream.LogPrefix = "iterateBadgerKVs"
+	stream.Send = func(kvl *pb.KVList) error {
+		kvs := make([]*pb.KV, 0, len(kvl.Kv))
+		for _, kv := range kvl.Kv {
+			if kv.Key != nil && kv.Value != nil {
+				kvs = append(kvs, kv)
+			}
+		}
+		if len(kvs) == 0 {
+			return nil
+		}
+		return iter(kvs)
+	}
+	return stream.Orchestrate(ctx)
+}
+
+func between(min, max, val int) int {
+	if val > max {
+		val = max
+	}
+	if val < min {
+		val = min
+	}
+	return val
+}
+
+func (b *BadgerV2) DefaultOptions(path string, readonly bool) Options {
+	var opts Options
+
+	bopts := badger.DefaultOptions(path)
+	bopts.DetectConflicts = false
+	bopts.CompactL0OnClose = true
+	bopts.Truncate = true
+	bopts.ValueLogLoadingMode = options.MemoryMap
+	bopts.TableLoadingMode = options.MemoryMap
+	bopts.ValueThreshold = 128
+	bopts.MaxTableSize = 64 << 20
+	bopts.ReadOnly = readonly
+
+	// Envvar LOTUS_CHAIN_BADGERSTORE_COMPACTIONWORKERNUM
+	if badgerNumCompactors, badgerNumCompactorsSet := os.LookupEnv("LOTUS_CHAIN_BADGERSTORE_COMPACTIONWORKERNUM"); badgerNumCompactorsSet {
+		if numWorkers, err := strconv.Atoi(badgerNumCompactors); err == nil && numWorkers >= 0 {
+			bopts.NumCompactors = numWorkers
+		}
+	}
+	opts.V2Options = bopts
+	return opts
+}
+
+func (b *BadgerV2) Load(r io.Reader, maxPendingWrites int) error {
+	return b.DB.Load(r, maxPendingWrites)
+}
+
+func (b *BadgerV2) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
+	return nil, fmt.Errorf("AllKeysChan is not implemented")
+}
+
+func (b *BadgerV2) DeleteBlock(context.Context, cid.Cid) error {
```

Comment:
I'm a bit confused about what's going on here.  Are these not implemented because they are never called?  If that's the case would it make more sense to just leave them out of the interface?

Reply:
Thanks - I think that these are not needed. I was stuck in mud metaphorically with the node builder system and may have temporarily had a configuration that tried to convert the badger object to a block object. I have removed DeleteBlock and AllKeysChan from the interface and the v2 and v4 implementation and optimistically as I write this now it appears everything is compiling properly and working as intended with them removed 4de5f718c123105dd98596b74c7dff302ff761af and 52604e948e28b80829f4fe131eec2a249aa40d26